File: examples/python/verify_organization.py
#!/usr/bin/env python3
"""
ApiPay.kz - Organization Verification Example

This example demonstrates the complete organization verification flow:
1. Start verification with IIN/BIN
2. Poll for status until verified or timeout

Prerequisites:
- Add phone 77056610934 to Kaspi Business as "Cashier"

Usage: API_KEY=your_key IDN=123456789012 python verify_organization.py
"""

import os
import requests
import sys
import time

API_KEY = os.environ.get('API_KEY')
IDN = os.environ.get('IDN')
API_BASE_URL = 'https://bpapi.bazarbay.site/api'

POLL_INTERVAL = 2  # seconds
MAX_TIMEOUT = 120  # seconds


def start_verification(idn: str) -> dict:
    """Start organization verification."""
    response = requests.post(
        f'{API_BASE_URL}/organizations/verify',
        headers={
            'X-API-Key': API_KEY,
            'Content-Type': 'application/json'
        },
        json={'idn': idn}
    )

    if not response.ok:
        error = response.json()
        raise Exception(f"API Error: {error.get('message', 'Unknown error')}")

    return response.json()


def check_status(organization_id: int) -> dict:
    """Check verification status."""
    response = requests.get(
        f'{API_BASE_URL}/organizations/{organization_id}/status',
        headers={'X-API-Key': API_KEY}
    )

    if not response.ok:
        error = response.json()
        raise Exception(f"API Error: {error.get('message', 'Unknown error')}")

    return response.json()


def wait_for_verification(organization_id: int) -> dict:
    """Poll for verification status until verified or timeout."""
    start_time = time.time()

    while time.time() - start_time < MAX_TIMEOUT:
        data = check_status(organization_id)
        organization = data['organization']
        status = organization['status']
        time_remaining = organization.get('time_remaining', 0)

        print(f'Status: {status}, Time remaining: {time_remaining}s')

        if status == 'verified':
            return organization

        if status == 'failed':
            raise Exception('Verification failed')

        time.sleep(POLL_INTERVAL)

    raise Exception('Verification timeout')


def main():
    if not API_KEY or not IDN:
        print('Error: API_KEY and IDN environment variables are required')
        print('Usage: API_KEY=your_key IDN=123456789012 python verify_organization.py')
        sys.exit(1)

    try:
        print('Starting organization verification...')
        print(f'IIN/BIN: {IDN}')
        print('')
        print('Please confirm in Kaspi Business app within 2 minutes.')
        print('---')

        result = start_verification(IDN)
        organization_id = result['organization']['id']

        print(f'Organization ID: {organization_id}')
        print('Waiting for confirmation...\n')

        organization = wait_for_verification(organization_id)

        print('\nVerification successful!')
        print('------------------------')
        print(f"Organization ID: {organization['id']}")
        print(f"IIN/BIN: {organization['idn']}")
        print(f"Status: {organization['status']}")
        print('\nYou can now create invoices!')

    except Exception as e:
        print(f'\nError: {e}')
        sys.exit(1)


if __name__ == '__main__':
    main()

File: examples/php/verify-organization.php
<?php
/**
 * ApiPay.kz - Organization Verification Example
 *
 * This example demonstrates the complete organization verification flow:
 * 1. Start verification with IIN/BIN
 * 2. Poll for status until verified or timeout
 *
 * Prerequisites:
 * - Add phone 77056610934 to Kaspi Business as "Cashier"
 *
 * Usage: API_KEY=your_key IDN=123456789012 php verify-organization.php
 */

$API_KEY = getenv('API_KEY');
$IDN = getenv('IDN');
$API_BASE_URL = 'https://bpapi.bazarbay.site/api';

$POLL_INTERVAL = 2;  // seconds
$MAX_TIMEOUT = 120;  // seconds

/**
 * Start organization verification
 */
function startVerification($idn) {
    global $API_KEY, $API_BASE_URL;

    $ch = curl_init("{$API_BASE_URL}/organizations/verify");
    curl_setopt_array($ch, [
        CURLOPT_POST => true,
        CURLOPT_HTTPHEADER => [
            "X-API-Key: {$API_KEY}",
            'Content-Type: application/json'
        ],
        CURLOPT_POSTFIELDS => json_encode(['idn' => $idn]),
        CURLOPT_RETURNTRANSFER => true
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);

    $data = json_decode($response, true);

    if ($httpCode >= 400) {
        throw new Exception("API Error: " . ($data['message'] ?? 'Unknown error'));
    }

    return $data;
}

/**
 * Check verification status
 */
function checkStatus($organizationId) {
    global $API_KEY, $API_BASE_URL;

    $ch = curl_init("{$API_BASE_URL}/organizations/{$organizationId}/status");
    curl_setopt_array($ch, [
        CURLOPT_HTTPHEADER => ["X-API-Key: {$API_KEY}"],
        CURLOPT_RETURNTRANSFER => true
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);

    $data = json_decode($response, true);

    if ($httpCode >= 400) {
        throw new Exception("API Error: " . ($data['message'] ?? 'Unknown error'));
    }

    return $data;
}

/**
 * Poll for verification status until verified or timeout
 */
function waitForVerification($organizationId) {
    global $POLL_INTERVAL, $MAX_TIMEOUT;

    $startTime = time();

    while (time() - $startTime < $MAX_TIMEOUT) {
        $data = checkStatus($organizationId);
        $organization = $data['organization'];
        $status = $organization['status'];
        $timeRemaining = $organization['time_remaining'] ?? 0;

        echo "Status: {$status}, Time remaining: {$timeRemaining}s\n";

        if ($status === 'verified') {
            return $organization;
        }

        if ($status === 'failed') {
            throw new Exception('Verification failed');
        }

        sleep($POLL_INTERVAL);
    }

    throw new Exception('Verification timeout');
}

// Main
if (!$API_KEY || !$IDN) {
    echo "Error: API_KEY and IDN environment variables are required\n";
    echo "Usage: API_KEY=your_key IDN=123456789012 php verify-organization.php\n";
    exit(1);
}

try {
    echo "Starting organization verification...\n";
    echo "IIN/BIN: {$IDN}\n\n";
    echo "Please confirm in Kaspi Business app within 2 minutes.\n";
    echo "---\n";

    $result = startVerification($IDN);
    $organizationId = $result['organization']['id'];

    echo "Organization ID: {$organizationId}\n";
    echo "Waiting for confirmation...\n\n";

    $organization = waitForVerification($organizationId);

    echo "\nVerification successful!\n";
    echo "------------------------\n";
    echo "Organization ID: {$organization['id']}\n";
    echo "IIN/BIN: {$organization['idn']}\n";
    echo "Status: {$organization['status']}\n";
    echo "\nYou can now create invoices!\n";

} catch (Exception $e) {
    echo "\nError: {$e->getMessage()}\n";
    exit(1);
}
